File: LCmac/pacificatlanticwaterflow.java
import java.util.*;;

public class pacificatlanticwaterflow {
    
}


class Solution {
    public List<List<Integer>> pacificAtlantic(int[][] heights) {

        if(heights == null){
            return null;

        }

        int rows = heights.length;
        int cols = heights[0].length;

        
        Set<List<Integer>> p= new HashSet<>();
        Set<List<Integer>> a = new HashSet<>();
        List<List<Integer>> res = new ArrayList<>();

        for(int r = 0; r<rows;r++){
            dfs(heights, r, 0, p, heights[r][0]);
            dfs(heights, r, cols - 1, a, heights[r][cols - 1]);
        }


        for(int c = 0; c<cols;c++){
            dfs(heights, 0, c, p, heights[0][c]);
            dfs(heights, rows-1, c, a, heights[rows-1][c]);
        }


        
        for(int r = 0;r<rows;r++){
            for(int c = 0;c<cols;c++){
                List<Integer> pos = Arrays.asList(r, c);
                if(p.contains(pos) && a.contains(pos)){
                    res.add(pos);



                }
            }
        }

        return res;



        
    }

    private void dfs(int[][] h, int r, int c, Set<List<Integer>> visit, int prevHeight){

        int rows = h.length;
        int cols = h[0].length;

        if (visit.contains(Arrays.asList(r, c)) || r < 0 || c < 0 || r >= rows || c >= cols || h[r][c] < prevHeight) {
            return;
        }
        visit.add(Arrays.asList(r, c));
        dfs(h, r + 1, c, visit, h[r][c]);
        dfs(h, r - 1, c, visit, h[r][c]);
        dfs(h, r, c + 1, visit, h[r][c]);
        dfs(h, r, c - 1, visit, h[r][c]);
    }
}

File: LCmac/ordered/graph/pacificatlanticwaterflow.py



directions = [[0,1], [0,-1], [-1,0], [1,0]]

pac = set()
atl = set()

rows, cols = len(heights), len(heights[0])

def dfs(r, c, visit, lastHeight):
    if (r,c) in visit or r<0 or c<0 or r>=rows or c>= cols or heights[r][c] < lastHeight:
        return 
    visit.add((r,c))

    for dr, dc in directions:
        dfs(r + dr, c + dc, visit, heights[r][c])

for r in range(rows):
    dfs(r, 0, pac, heights[r][0])
    dfs(r, cols -1, atl, heights[r][cols-1])

for c in range(cols):
    dfs(0, c, pac, heights[0][c])
    dfs(rows-1, c, atl, heights[rows-1][c])

res = pac & atl
return [[r.c] for (r,c) in res]

